osquery/carver/carver.cpp

```diff
@@ -0,0 +1,311 @@
+/*
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#ifdef WIN32
+#define _WIN32_DCOM
+#define WIN32_LEAN_AND_MEAN
+#include <Windows.h>
+#endif
+
+#include <boost/property_tree/ptree.hpp>
+
+// This define is required for Windows static linking of libarchive
+#define LIBARCHIVE_STATIC
+#include <archive.h>
+#include <archive_entry.h>
+
+#include <osquery/filesystem.h>
+#include <osquery/flags.h>
+#include <osquery/logger.h>
+
+#include "osquery/carver/carver.h"
+#include "osquery/core/conversions.h"
+#include "osquery/core/json.h"
+#include "osquery/filesystem/fileops.h"
+#include "osquery/remote/requests.h"
+#include "osquery/remote/serializers/json.h"
+#include "osquery/remote/transports/tls.h"
+#include "osquery/remote/utility.h"
+#include "osquery/tables/system/hash.h"
+
+namespace fs = boost::filesystem;
+namespace pt = boost::property_tree;
+
+namespace osquery {
+
+/// Prefix used for the temp FS where carved files are stored
+const std::string kCarvePathPrefix = "osquery_carve_";
+
+/// Prefix applied to the file carve tar archive.
+const std::string kCarveNamePrefix = "carve_";
+
+/// Database prefix used to directly access and manipulate our carver entries
+const std::string kCarverDBPrefix = "carves.";
+
+/*
+ * Carver block chunking size. This specifies how big of a block size to
+ * use when carving a file from disk.
+ */
+const size_t kBuffSize = 8192;
+
+DECLARE_string(tls_hostname);
+
+/// Session creation endpoint for forensic file carve
+CLI_FLAG(string,
+         carver_start_endpoint,
+         "",
+         "TLS/HTTPS init endpoint for forensic carver");
+
+/// Data aggregation endpoint for forensic file carve
+CLI_FLAG(
+    string,
+    carver_continue_endpoint,
+    "",
+    "TLS/HTTPS endpoint that receives carved content after session creation");
+
+/// Size of blocks used for POSTing data back to remote endpoints
+CLI_FLAG(int32,
+         carver_block_size,
+         8192,
+         "Size of blocks used for POSTing data back to remote endpoints");
+
+CLI_FLAG(bool,
+         disable_carver,
+         true,
+         "Disable the osquery file carver (default true)");
+
+/// Helper function to update values related to a carve
+void updateCarveValue(const std::string& guid,
+                      const std::string& key,
+                      const std::string& value) {
+  std::string carve;
+  auto s = getDatabaseValue(kQueries, kCarverDBPrefix + guid, carve);
```

Comment:
Don't use `kQueries`, perhaps there's another domain we can add, or use the general persistent domain.